cumulusci/tasks/apex/testrunner.py

```diff
@@ -222,15 +230,36 @@ def _get_test_classes(self):
         self.logger.info("Found {} test classes".format(result["totalSize"]))
         return result
 
-    def _get_test_results(self):
+    def _is_retriable_failure(self, test_result):
+        return test_result["Outcome"] == "Fail" and any(
+            [
+                reg.search(test_result["Message"] or "")
+                or reg.search(test_result["StackTrace"] or "")
+                for reg in self.options["retry_errors"]
+            ]
+        )
+
+    def _get_test_results(self, allow_retries=True):
         result = self.tooling.query_all(TEST_RESULT_QUERY.format(self.job_id))
-        self.counts = {"Pass": 0, "Fail": 0, "CompileFail": 0, "Skip": 0}
+
+        if allow_retries:
+            self.retry_details = {}
+
         for test_result in result["records"]:
             class_name = self.classes_by_id[test_result["ApexClassId"]]
             self.results_by_class_name[class_name][
                 test_result["MethodName"]
             ] = test_result
             self.counts[test_result["Outcome"]] += 1
+
+            # Determine whether this failure is retriable.
+            if allow_retries and self._is_retriable_failure(test_result):
+                self.counts["Retriable"] += 1
+                self.retry_details.setdefault(test_result["ApexClassId"], []).append(
+                    test_result["MethodName"]
+                )
```

Comment:
Consider initializing retry_details as defaultdict(list) and then this can just be `self.retry_details[test_result["ApexClassId"]].append(test_result["MethodName"])`